### ISSUE TITLE
Support for entity regions

### ISSUE THREAD

**Issue by erlend-sh** [Dooskington/grubbnet#1]
In your [[20] Milestone](https://ratwizard.dev/dev-log/antorum/20) dev log (all of which is amazing btw!) you share some detail on the following feature:

> I have been waiting a long time to add some concept of "entity regions" into the game, and having that done basically means that the entire game world is no longer regarded as one large room with all players within it. As entities move around the island, they are discretely sorted and swapped between regions, syncing themselves (or the removal of themselves) to any players that need it. Players that are in an entirely different region than some other entity are no longer required to keep track of it, and will no longer be sent any packets it may generate. Only the entities relevant to a client are streamed to that client.

This is quite novel. I know Red Dead Redemption 2 does something similar for its online play.

Might you share this as part of grubbnet or a separate utility crate?

**Comment by Dooskington** [comment]
Hello! The entity region code is closer to game code than it is to networking code, so I don't think it would fit into the `grubbnet` crate. However, I hadn't thought about placing it in a separate utility crate - that is a good idea. 

As it stands, I was planning on improving some aspects of the system soon, mostly to stomp out some bugs. When I do that, I'll investigate putting it into a shared crate. Thanks for the suggestion!

**Comment by erlend-sh** [comment]
Great!

Hope you don’t mind me going a little bit off topic (seems your email is not public): If you’re looking for more ideas for the dev log, I’d love to know more about your Rust-backend + Unity-frontend setup. I know several Rust games that do this with Godot already, but this is the first serious attempt I’m seeing with Unity. I’m not very familiar with how well they support this decoupled approach, so more details on that would be highly informative.

A year ago in [8] you wrote:

> If Unity native plugins weren't such a hassle to work with, maybe I would have written the terrain engine in Rust so I could also share it with the server.

That turned out not to be a difficult constraint to live with the first time you encountered it. I’m curious if you’ve run into something similar again, or if the decoupled setup is largely painless.

I suppose you’re using Unity instead of Godot because it just has more built-in utilities & add-ons for rapid development?

- - -

What’s the best way to make comments on the devlog? I’ve got some more questions & thoughts I’d like to follow up with, just not sure where to put it.

**Comment by Dooskington** [comment]
I have been wondering if it's worth doing a more technical post. I can for sure try and give more details about my client / server setup in such a post. It has evolved a good bit since the project first started, and it really is far from perfect. I was planning on entirely rewriting the pipeline once the game is finished and I have a 100% complete list of requirements (and start working on the next project).

I can try to answer some of your questions here about the setup,
The server has the entire game codebase: all logic, resources (items, drop tables, entities), pathfinding, database stuff, etc.
It uses `specs` entity component system. Networked components are synced across the network and replicated on the client as unity components (which I write separately, there is no cool code-generation here). Honestly the entity syncing (both network and database) and how it integrates with the entity component system is a tricky topic and too deep to get into here.

On the client-side I have written a networking layer in C# that can connect & send/receive to the game server (Basically it's a clientside `grubbnet` implementation for Unity). So I've got a well-defined protocol of game updates / state changes / remote procedure calls that the server generates, and the client knows how to read them. From that point on, the puzzle is to replicate everything within the bounds of Unity. So, a server-side entity instantiate needs to generate a packet which eventually gets interpreted on the client as a GameObject Instantiate, and then those objects need to be linked via network id.

> That turned out not to be a difficult constraint to live with the first time you encountered it. I’m curious if you’ve run into something similar again, or if the decoupled setup is largely painless.

The biggest bit of shared data is the map (just the terrain - clients are given a map with no entity data), which just follows a map file specification that has grown over time with the project. Since both the server and client can read it, it is trivial to perform validation/collision on the server, while replicating it visually on the client. So, because of that, having the terrain engine being exclusive to the client-side (the game and editor) was not such a big deal. It may have been nice to have a shared implementation as a native library, but at this point in the project I can say it would have taken a lot more time for little benefit. A lot of the terrain engine code is about mesh generation and unity specific gameobject manipulation, so the server doesn't need to care about it.

> I suppose you’re using Unity instead of Godot because it just has more built-in utilities & add-ons for rapid development?

I'm using Unity instead of Godot because I've just been using it for a really long time and I am super comfortable with it. I have never tried Godot before but it looks really promising. I am sure it would work well (maybe even better) for a project like this!

Honestly I'd love to write another Antorum client in a different language / engine, just for fun. Also, half the point of this project is that somebody will try and write their own client, given the open protocol :)

> What’s the best way to make comments on the devlog? I’ve got some more questions & thoughts I’d like to follow up with, just not sure where to put it.

You inspired me, so I'm adding a comments section to the dev logs. It will hopefully be available in the next couple days :)
Feel free to post your follow up questions on future or past dev logs.

**Comment by erlend-sh** [comment]
> Also, half the point of this project is that somebody will try and write their own client, given the open protocol :)

Does this open protocol already exist or are you talking about future plans here?

**Comment by Dooskington** [comment]
Unfortunately the documentation for the protocol is a bit messy right now, and is changing rapidly as I iterate on core bits of the game. However, I am planning on releasing the initial protocol specs along with the first playable build of the game, hopefully some time soon.
